### PR TITLE
AESKW support

### PR DESCRIPTION
**Pull request by ifranzki**
Add support for the `CKM_RSA_AES_KEY_WRAP` mechanism in the CCA, ICA and Soft tokens. For the ICA and Soft tokens also support for mechanisms `CKM_AES_KEY_WRAP`, `CKM_AES_KEY_WRAP_PAD`, `CKM_AES_KEY_WRAP_KWP`, `CKM_AES_KEY_WRAP_PKCS7`, and `CKM_ECDH_AES_KEY_WRAP` is added.

### CODE REVIEW

File: COPYRIGHTS [hunk 1]
```diff
@@ -33,6 +33,10 @@ For code originating from OpenSSL:
  *   RSA_padding_check_PKCS1_OAEP_mgf1() in crypto/rsa/rsa_oaep.c and is
  *   slightly modified to fit to the OpenCryptoki environment. See comment in
  *   function decode_eme_oaep() for a list of changes.
+ * - The implementation of the AESKW functions in usr/ib/common/mech_aes.c is
```
Comment (jschmidb) [inline COPYRIGHTS:36]: usr/lib .... also below

File: usr/lib/common/mech_aes.c [hunk 2]
```diff
@@ -4328,3 +4328,637 @@ CK_RV aes_xts_cipher(CK_BYTE *in_data, CK_ULONG in_data_len,
...
+
+    if (*out_data_len < padded_len + AES_KEY_WRAP_BLOCK_SIZE) {
+        TRACE_ERROR("%s\n", ock_err(ERR_BUFFER_TOO_SMALL));
+        return CKR_BUFFER_TOO_SMALL;
+    }
+
+    if (iv == NULL)
+        memcpy(aiv, aeskw_default_aiv, AES_KEY_WRAP_KWP_IV_SIZE);
+    else
+        memcpy(aiv, iv,AES_KEY_WRAP_KWP_IV_SIZE);
```
Comment (jschmidb) [inline usr/lib/common/mech_aes.c:4544]: space after iv,  ?

File: COPYRIGHTS (hunk 1 above)
Comment (ifranzki) [reply COPYRIGHTS:36]: fixed

File: testcases/crypto/aes.h [hunk 3]
```diff
@@ -1058,6 +1061,123 @@ static struct aes_test_vector aes_xts_tv[] = {
...
+        .klen = 24,
+        .plaintext = {0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77,
+                      0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF,},
+        .plen = 16,
+        .ciphertext = {0x96, 0x77, 0x8B, 0x25, 0xAE, 0x6C, 0xA4, 0x35,
+                       0xF9, 0x2B, 0x5B, 0x97, 0xC0, 0x50, 0xAE, 0xD2,
+                       0x46, 0x8A, 0xB8, 0xA1, 0x7A, 0xD8, 0x4E, 0x5D,},
+        .clen = 24,
+    },
+    {                           // #2
```
Comment (jschmidb) [inline testcases/crypto/aes.h:1090]: move }, { to one line as above?  (same below)

Comment (ifranzki) [reply testcases/crypto/aes.h:1090]: fixed

File: usr/lib/common/key_mgr.c [hunk 4]
```diff
@@ -1424,6 +1427,18 @@ CK_RV key_mgr_unwrap_key(STDLL_TokData_t *tokdata,
     else
         fromend = FALSE;
 
+    if (mech->mechanism == CKM_RSA_AES_KEY_WRAP) {
+        /*
+         * For CKM_RSA_AES_KEY_WRAP, the mechanism parameter copy in ctx has
+         * been updated to hold the actual size of the temp. AES key. Use this
+         * for statistics counting, not the user supplied mechanism parameter,
+         * which might contain an invalid/incorrect AES key size.
+         * Ctx is cleanedup and freed only after statistics counting, so the
```
Comment (jschmidb) [inline usr/lib/common/key_mgr.c:1439]: cleaned up

File: usr/lib/common/mech_rsa.c [hunk 5]
```diff
@@ -2997,3 +2998,282 @@ CK_RV check_pss_params(CK_MECHANISM *mech, CK_ULONG modlen)
...
+    }
+
+    /* Generate a temporary AES key of the desired size */
+    aes_key_size = params->ulAESKeyBits / 8;
+    rc = key_mgr_generate_key(tokdata, sess, &aes_keygen_mech,
+                              aes_key_tmpl,
+                              sizeof(aes_key_tmpl) / sizeof(CK_ATTRIBUTE),
+                              &aes_key_handle, FALSE);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to genrate temporary AES key (%lu bits): "
```
Comment (jschmidb) [inline usr/lib/common/mech_rsa.c:3048]: generate

File: usr/lib/common/mech_rsa.c [hunk 6]
```diff
@@ -2997,3 +2998,282 @@ CK_RV check_pss_params(CK_MECHANISM *mech, CK_ULONG modlen)
...
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("Failed to get CKA_VALUE_LEN from temp. AES key.\n");
+        return rc;
+    }
+
+    /* Update the mechanism param (copy) in context. Might be needed later on */
+    params->ulAESKeyBits = value_len * 8;
+
+    /*
+     * /Decrypt (unwrap) the final key data from the the second part of the
```
Comment (jschmidb) [inline usr/lib/common/mech_rsa.c:3245]: remove leading /

File: usr/lib/cca_stdll/cca_specific.c [hunk 7]
```diff
@@ -4366,6 +4366,204 @@ static CK_RV cca_get_min_card_level(STDLL_TokData_t *tokdata)
...
+static CK_RV cca_get_acp_info_handler(STDLL_TokData_t *tokdata,
+                                      const char *adapter,
+                                      unsigned short card,
+                                      unsigned short domain,
+                                      void *handler_data)
+{
+    cca_acp_info_data_t *data = (cca_acp_info_data_t *)handler_data;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0, };
+    unsigned char role_name[8] = { 0 };
+    unsigned char roles_data[4000], role_data[4000];
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:4436]: intentionally no init to zeros for roles_data and role_data?

File: usr/lib/cca_stdll/cca_specific.c [hunk 8]
```diff
@@ -4366,6 +4366,204 @@ static CK_RV cca_get_min_card_level(STDLL_TokData_t *tokdata)
...
+    long return_code, reason_code, rule_array_count, roles_data_len;
+    long i, role_data_len;
+    struct cca_role_data *role_info;
+    CK_BBOOL found = CK_FALSE;
+
+    UNUSED(tokdata);
+    UNUSED(adapter);
+
+    TRACE_DEBUG("APQN %02X.%04X (adapter '%s')\n", card, domain, adapter);
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:4445]: adapter is used here, but flagged as UNUSED above

File: usr/lib/cca_stdll/cca_stdll.h [hunk 9]
```diff
@@ -208,6 +208,34 @@ struct cca_key_derivation_data {
...
+    uint16_t       auth_level;
+    uint16_t       lower_time_limit;   /* 0xhhmm */
+    uint16_t       upper_time_limit;   /* 0xhhmm */
+    uint8_t        days_valid;         /* 7 bits, Sunday, Monday ... Saturday */
+    uint16_t       num_segments;
+    uint16_t       reserved;
+} __attribute__ ((__packed__));
+
+struct cca_acp_info {
+    CK_BBOOL acp_03B8; /* Symmetric Key Export - AES, CKM-RAKW */
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_stdll.h:234]: CKM-RAKW ? is this a mechanism?

File: usr/lib/cca_stdll/cca_specific.c [hunk 10]
```diff
@@ -13160,6 +13237,418 @@ static CK_RV ccatok_unwrap_key_rsa_pkcs(STDLL_TokData_t *tokdata,
...
+    }
+
+    if (keytype != sec_aes_cipher_key) {
+        TRACE_ERROR("CCA does not support wrapping with CKM_RSA_AES_KEY_WRAP "
+                    "for AES DATA keys\n");
+        return CKR_KEY_NOT_WRAPPABLE;
+    }
+
+    rule_array_count = 2;
+    memcpy(rule_array, "AES     CKM-RAKW", 2 * CCA_KEYWORD_SIZE);
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:13275]: ah, got it ... it's a CCA rule array keyword

File: usr/lib/cca_stdll/cca_specific.c [hunk 11]
```diff
@@ -13160,6 +13237,418 @@ static CK_RV ccatok_unwrap_key_rsa_pkcs(STDLL_TokData_t *tokdata,
...
+    long return_code, reason_code, rule_array_count;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0 };
+    CK_BYTE buffer[3500] = { 0, };
+    long buffer_len = sizeof(buffer);
+    enum cca_token_type keytype;
+    unsigned int keybitsize;
+    const CK_BYTE *mkvp;
+
+    if (!cca_rsa_aeskw_supported(tokdata, CKK_AES)) {
+        TRACE_ERROR("CKM_RSA_AES_KEY_WRAP requires CCA 7.5 or later and "
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:13256]: CCA 8.2 ?

File: usr/lib/cca_stdll/cca_specific.c [hunk 12]
```diff
@@ -13160,6 +13237,418 @@ static CK_RV ccatok_unwrap_key_rsa_pkcs(STDLL_TokData_t *tokdata,
...
+    CK_ULONG key_size = 0;
+    enum cca_token_type keytype;
+    unsigned int keybitsize;
+    const CK_BYTE *mkvp;
+    CK_BBOOL new_mk;
+    CK_IBM_CCA_AES_KEY_MODE_TYPE mode;
+    CK_RV rc;
+
+    if (!cca_rsa_aeskw_supported(tokdata, CKK_AES)) {
+        TRACE_ERROR("CKM_RSA_AES_KEY_WRAP requires CCA 7.5 or later and "
```
Comment (jschmidb) [inline usr/lib/cca_stdll/cca_specific.c:13424]: CCA 8.2 ?

File: usr/lib/common/mech_ec.c [hunk 13]
```diff
@@ -1645,3 +1646,403 @@ int ec_point_uncompressed_from_public_data(const CK_BYTE *data,
...
+    ec_publ_key_tmpl[0] = *ec_params;
+    rc = key_mgr_generate_key_pair(tokdata, sess, &ec_keygen_mech,
+                                   ec_publ_key_tmpl, sizeof(ec_publ_key_tmpl) /
+                                                       sizeof(CK_ATTRIBUTE),
+                                   ec_priv_key_tmpl, sizeof(ec_priv_key_tmpl) /
+                                                       sizeof(CK_ATTRIBUTE),
+                                   &ec_publ_key_handle, &ec_priv_key_handle,
+                                   FALSE);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to genrate temporary EC key pair: "
```
Comment (jschmidb) [inline usr/lib/common/mech_ec.c:1744]: generate

File: usr/lib/common/mech_ec.c [hunk 14]
```diff
@@ -1645,3 +1646,403 @@ int ec_point_uncompressed_from_public_data(const CK_BYTE *data,
...
+                            sizeof(aes_key_tmpl) / sizeof(CK_ATTRIBUTE),
+                            FALSE);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to derive temporary AES key (%lu bits): "
+                    "%s (0x%lx)\n", params->ulAESKeyBits, p11_get_ckr(rc), rc);
+        goto done;
+    }
+
+    /*
+     * /Decrypt (unwrap) the final key data from the the second part of the
```
Comment (jschmidb) [inline usr/lib/common/mech_ec.c:2010]: Leading /

File: usr/lib/common/key_mgr.c (hunk 4 above)
Comment (ifranzki) [reply usr/lib/common/key_mgr.c:1439]: fixed

File: usr/lib/common/mech_rsa.c (hunk 5 above)
Comment (ifranzki) [reply usr/lib/common/mech_rsa.c:3048]: fixed

File: usr/lib/common/mech_rsa.c (hunk 6 above)
Comment (ifranzki) [reply usr/lib/common/mech_rsa.c:3245]: fixed

File: usr/lib/cca_stdll/cca_specific.c (hunk 7 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:4436]: yes, those will be set by the CCA verb. No need to initialize them.

File: usr/lib/cca_stdll/cca_specific.c (hunk 8 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:4445]: Yes, but it trace is off, then the TRACE_DEBUG macro won't use the parameter, and then it is unused. Other than in this trace stmt, `adapter` is not used in this function. It does not hurt to declare a variable as unused even if it is used.

I could enclose the `UNUSED(adapter)` in a `#ifndef DEBUG` block, but I think this does not make it easier to understand either....

File: usr/lib/cca_stdll/cca_specific.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:13256]: fixed

Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:13256]: fixed

File: usr/lib/cca_stdll/cca_specific.c (hunk 12 above)
Comment (ifranzki) [reply usr/lib/cca_stdll/cca_specific.c:13424]: fixed

File: usr/lib/common/mech_ec.c (hunk 13 above)
Comment (ifranzki) [reply usr/lib/common/mech_ec.c:1744]: fixed

File: usr/lib/common/mech_ec.c (hunk 14 above)
Comment (ifranzki) [reply usr/lib/common/mech_ec.c:2010]: fixed

**Review by jschmidb — APPROVED** [review]
lgtm

**Pull request merged**